### PR TITLE
Fix plugins install not returning correct exit code on failure

### LINKED ISSUE #239: Custom image build reports success on a failed plugin installation
When trying to build a custom BitOps image with the plugin referring to the wrong git repo, the image build succeeds while in fact there was a failure.

Build the custom "plugins" image with the following `bitops.config.yaml`:
```
bitops:
  fail_fast: true     # When set, will exit if any warning+ is raised, otherwise only exit on critical error
  logging:      
    level: DEBUG
    color:
      enabled: true
    filename: bitops-run
    err: bitops.logs
    path: /var/logs/bitops
  opsrepo_root_default_dir: _default
  plugins:
    helm:
      source: https://google.com/ # <--- wrong git repo
  deployments:
    helm:
      plugin: helm
```

Let's take a look at what's inside the image: 
```sh
docker run -it custom/my-bitops:helm --entrypoint /bin/sh
```

Re-running the original `plugins.py install` command:
```sh
/opt/bitops # python3 scripts/plugins.py install
2022-08-01 16:52:47,564 git.cmd      DEBUG Popen(['git', 'version'], cwd=/opt/bitops, universal_newlines=False, shell=None, istream=None)
2022-08-01 16:52:47,569 git.cmd      DEBUG Popen(['git', 'version'], cwd=/opt/bitops, universal_newlines=False, shell=None, istream=None)
2022-08-01 16:52:47,641 root         INFO 

 ____  _ _    ___            
| __ )(_) |_ / _ \ _ __  ___ 
|  _ \| | __| | | | '_ \/ __|
| |_) | | |_| |_| | |_) \__ \
|____/|_|\__|\___/| .__/|___/
                  |_|        

    BitOps is a way to describe the infrastructure and things deployed onto that infrastructure for multiple environments in a single place called an Operations Repo.
    
2022-08-01 16:52:47,641 root         INFO 


#~#~#~#~ BITOPS CONFIGURATION ~#~#~#~        
        FAIL FAST:              [True]                       
        RUN MODE:               [default]                        
        DEFAULT RUN MODE:       [install]                            
        LOGGING LEVEL:          [DEBUG]                       
        LOGGING COLOR:          [True]                           
        BITOPS CONFIG FILE:     [bitops.config.yaml]                     
#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#
                       
2022-08-01 16:52:47,654 root         INFO 


~#~#~#~PROCESSING STAGE [HELM]~#~#~#~

2022-08-01 16:52:47,655 root         INFO 


~#~#~#~PLUGIN SOURCE [https://google.com/]~#~#~#~

2022-08-01 16:52:47,657 root         INFO 


~#~#~#~CLONING PLUGIN [HELM]~#~#~#~              
         PLUGIN_SOURCE:         [https://google.com/]                           
         PLUGIN_TAG:            [latest]                              
         PLUGIN_BRANCH:         [main]                            
#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~#~# 
                            
2022-08-01 16:52:47,660 git.cmd      DEBUG Popen(['git', 'clone', '-v', '--branch=main', 'https://google.com/', '/opt/bitops/scripts/plugins/helm'], cwd=/opt/bitops, universal_newlines=True, shell=None, istream=None)
2022-08-01 16:52:48,040 git.repo.base DEBUG Cmd(['git', 'clone', '-v', '--branch=main', 'https://google.com/', '/opt/bitops/scripts/plugins/helm'])'s unused stdout: 
2022-08-01 16:52:48,041 git.cmd      DEBUG AutoInterrupt wait stderr: b"Cloning into '/opt/bitops/scripts/plugins/helm'...\nfatal: repository 'https://google.com/' not found\n"
2022-08-01 16:52:48,043 root         INFO 
~#~#~#~CLONING PLUGIN [helm] FAILED~#~#~#~
        Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v --branch=main https://google.com/ /opt/bitops/scripts/plugins/helm
  stderr: 'Cloning into '/opt/bitops/scripts/plugins/helm'...
fatal: repository 'https://google.com/' not found
'
```

It failed, bit looks like the script didn't report the correct return code to be considered a failure:
```
/opt/bitops # echo $?
0
```

### PR DESCRIPTION
**Pull request by armab**
Fixes #239 by adding error exit codes where plugin installation has failed.

### CODE REVIEW

File: scripts/plugins/install_plugins.py [hunk 1]
```diff
@@ -66,20 +68,20 @@ def Install_Plugins():
...
 
             except git.exc.GitCommandError as exc:
-                logger.info("\n~#~#~#~CLONING PLUGIN [{plugin_config}] FAILED~#~#~#~\n\t{stderr}"
+                logger.error("\n~#~#~#~CLONING PLUGIN [{plugin_config}] FAILED~#~#~#~\n\t{stderr}"
                     .format(
                         plugin_config=plugin_config,
                         stderr=exc
                     ))
-                if BITOPS_fast_fail_mode: quit()
+                sys.exit(1)
```
Comment (armab) [inline scripts/plugins/install_plugins.py:76]: Looks like the `fail_fast` is not really implemented, but it's outside of this PR.

**Comment by PhillypHenning** [comment]
What's here is fine for 2.0.0, however, for 2.0.1 we will be looking to update all error messages to use the `documentation.json` + doc lookup functionality. 

This will keep a consistent and maintainable pattern for error messages and codes.

Example;

Utilities.py (Today)
```
except git.exc.GitCommandError as exc:
                logger.error("\n~#~#~#~CLONING PLUGIN [{plugin_config}] FAILED~#~#~#~\n\t{stderr}"
                    .format(
                        plugin_config=plugin_config,
                        stderr=exc
                    ))
                sys.exit(1) <-- FINE FOR TODAY
```


Utilities.py (Tomorrow)
```
from doc import Get_Doc

except git.exc.GitCommandError as exc:
                logger.error("\n~#~#~#~CLONING PLUGIN [{plugin_config}] FAILED~#~#~#~\n\t{stderr}"
                    .format(
                        plugin_config=plugin_config,
                        stderr=exc
                    ))
                sys.exit( Get_Doc("git_clone_error") ) <-- Intention of tomorrow
```


Documentation.json
```
{
    "missing_plugin_dependency": {
        "msg": "Missing plugin dependency. To Fix this update the bitops/bitops.config.yaml file by adding the dependent into the bitops.plugins block.",
        "link": ""
    },
    "github_clone_error" : {
        "msg": "Github repo failed to clone",
         "link": "url-to-doc"
         "code": "501"
}
```

**Comment by armab** [comment]
The possible range for exit codes is between 1 and 255, https://unix.stackexchange.com/questions/418784/what-is-the-min-and-max-values-of-exit-codes-in-linux

The question would be how much value it can provide for the users to have a unique code for each error and the overhead for us keeping that in a dedicated .json. How many would really care if the plugin install failed with `130` or `140` and how many will lookup that in the docs when the message was already printed to them?

Could be a topic for another discussion, but I would focus on `success` vs `failure` for now.

**Comment by armab** [comment]
We can follow some known codes and conventions where possible though, check:  https://stackoverflow.com/a/1101969/4533625

**Review by PhillypHenning — APPROVED** [review]

File: scripts/plugins/install_plugins.py (hunk 1 above)
Comment (armab) [reply scripts/plugins/install_plugins.py:76]: If we're going to release v2.0.0 that advertises `fail_fast` functionality (yaml setting, docs https://github.com/bitovi/bitops/blob/plugins/docs/plugins.md#bitopsconfigyaml), but doesn't work yet, - it perhaps better to remove it for now until the full implementation.

How about removing it just from the docs for now so it doesn't break the expectations?  (hidden)

**Pull request merged**